File: experiments-local/experiment-16/runs/agent=omp_language=typescript_model=openrouter/moonshotai/kimi-k2.7-code_tooling=none/rep1/src/normalize.ts
/**
 * Normalization helpers for team names, dates, and numeric values.
 *
 * Brazilian soccer datasets are messy: state suffixes ("-SP"), full legal names,
 * accented characters, and multiple date formats all appear. This module
 * centralizes the cleanup so queries and data rows speak the same language.
 */

import { createHash } from "node:crypto";

const STOPWORDS = new Set([
  "futebol",
  "clube",
  "esporte",
  "club",
  "sociedade",
  "esportiva",
  "associação",
  "associacao",
  "desportiva",
  "desportos",
  "esportivo",
  "football",
]);

const STATE_SUFFIX_RE = /\s*-\s*([A-Za-z]{2})\s*$/;
const PARENS_RE = /\s*\([^)]*\)\s*$/;

function removeAccents(input: string): string {
  return input.normalize("NFD").replace(/[\u0300-\u036f]/g, "");
}

/**
 * Turn a raw team name into normalized searchable text.
 *
 * Hyphenated state codes ("Flamengo-RJ") are converted to a trailing word
 * ("flamengo rj") so aliases can match both suffixed and unsuffixed forms.
 */
function cleanTokens(input: string): string {
  const lowered = removeAccents(input).toLowerCase();
  const withSpacedState = lowered.replace(STATE_SUFFIX_RE, " $1");
  const withoutParens = withSpacedState.replace(PARENS_RE, "");
  return withoutParens;
}

function tokenKey(input: string): string {
  return cleanTokens(input)
    .split(/\s+/)
    .filter((t) => t.length > 0 && !STOPWORDS.has(t))
    .sort()
    .join(" ");
}

interface ClubEntry {
  display: string;
  aliases: string[];
}

const CLUB_ALIASES: Record<string, ClubEntry> = {
  flamengo: { display: "Flamengo", aliases: ["flamengo", "crf", "clube de regatas do flamengo", "flamengo rj"] },
  palmeiras: { display: "Palmeiras", aliases: ["palmeiras", "sep", "sociedade esportiva palmeiras", "palmeiras sp"] },
  corinthians: { display: "Corinthians", aliases: ["corinthians", "sccp", "sport club corinthians paulista", "corinthians sp"] },
  "sao paulo": { display: "São Paulo", aliases: ["sao paulo", "spfc", "sao paulo fc", "são paulo", "sao paulo sp"] },
  santos: { display: "Santos", aliases: ["santos", "sfc", "santos sp"] },
  gremio: { display: "Grêmio", aliases: ["gremio", "gremio fbpa", "gremio rs"] },
  internacional: { display: "Internacional", aliases: ["internacional", "inter", "sport club internacional", "internacional rs"] },
  cruzeiro: { display: "Cruzeiro", aliases: ["cruzeiro", "cruzeiro esporte clube", "cruzeiro mg"] },
  "atletico mineiro": { display: "Atlético Mineiro", aliases: ["atletico mineiro", "atletico mg", "athletico mineiro", "galo", "clube atletico mineiro", "atletico mg"] },
  "atletico paranaense": { display: "Athletico Paranaense", aliases: ["atletico paranaense", "athletico paranaense", "atletico pr", "athletico pr", "furacao", "clube atletico paranaense", "atletico pr"] },
  fluminense: { display: "Fluminense", aliases: ["fluminense", "flu", "fluminense football club", "fluminense rj"] },
  botafogo: { display: "Botafogo", aliases: ["botafogo", "botafogo rj", "botafogo de futebol e regatas"] },
  vasco: { display: "Vasco da Gama", aliases: ["vasco", "vasco da gama", "vasco rj", "club de regatas vasco da gama", "vasco da gama rj"] },
  bahia: { display: "Bahia", aliases: ["bahia", "esporte clube bahia", "bahia ec", "ec bahia", "bahia ba"] },
  fortaleza: { display: "Fortaleza", aliases: ["fortaleza", "fortaleza esporte clube", "fortaleza fc", "fortaleza ce"] },
  ceara: { display: "Ceará", aliases: ["ceara", "ceara sporting club", "ceará", "ceara ce"] },
  sport: { display: "Sport Recife", aliases: ["sport", "sport recife", "sport club do recife", "sport pe"] },
  vitoria: { display: "Vitória", aliases: ["vitoria", "vitoria esporte clube", "vitória", "vitoria ba"] },
  goias: { display: "Goiás", aliases: ["goias", "goias esporte clube", "goiás", "goias go"] },
  coritiba: { display: "Coritiba", aliases: ["coritiba", "coritiba foot ball club", "coritiba pr"] },
  "ponte preta": { display: "Ponte Preta", aliases: ["ponte preta", "associacao atletica ponte preta", "ponte preta sp"] },
  nautico: { display: "Náutico", aliases: ["nautico", "nautico capibaribe", "náutico", "nautico pe"] },
  "america mineiro": { display: "América Mineiro", aliases: ["america mineiro", "america mg", "américa mineiro", "américa mg"] },
  chapecoense: { display: "Chapecoense", aliases: ["chapecoense", "associacao chapecoense de futebol", "chapecoense sc"] },
  avai: { display: "Avaí", aliases: ["avai", "avai fc", "avaí", "avai sc"] },
};

const ALIAS_TO_CANONICAL: Map<string, string> = new Map();
for (const [canonical, entry] of Object.entries(CLUB_ALIASES)) {
  ALIAS_TO_CANONICAL.set(canonical, canonical);
  for (const alias of entry.aliases) {
    ALIAS_TO_CANONICAL.set(tokenKey(alias), canonical);
  }
}

/**
 * Return the canonical key for a team name. If the name matches a known alias,
 * the alias map is used; otherwise a normalized token key is returned.
 */
export function teamKey(name: string | null | undefined): string {
  if (!name) return "";
  const trimmed = name.trim();

  // Try the full name with hyphenated state codes converted to words.
  const direct = ALIAS_TO_CANONICAL.get(tokenKey(trimmed));
  if (direct) return direct;

  // Fall back to the base name with the state suffix removed entirely.
  const base = trimmed.replace(STATE_SUFFIX_RE, "").trim();
  const baseAlias = ALIAS_TO_CANONICAL.get(tokenKey(base));
  if (baseAlias) return baseAlias;

  return tokenKey(base);
}

/**
 * Return a display name for a team key. Falls back to a cleaned version of the
 * raw name when no canonical alias exists.
 */
export function teamDisplay(name: string | null | undefined, key?: string): string {
  if (!name) return key ?? "Unknown";
  const k = key ?? teamKey(name);
  const canonical = CLUB_ALIASES[k];
  if (canonical) return canonical.display;
  let cleaned = cleanTokens(name)
    .replace(PARENS_RE, "")
    .replace(STATE_SUFFIX_RE, "")
    .replace(/\s+[a-z]{2}\s*$/, "")
    .trim();
  if (!cleaned) cleaned = name.trim();
  return cleaned
    .split(" ")
    .map((w) => w.charAt(0).toUpperCase() + w.slice(1).toLowerCase())
    .join(" ");
}

/**
 * Normalize a competition name to a small set of categories.
 */
export function normalizeCompetition(name: string): string {
  const lower = removeAccents(name).toLowerCase().trim();
  if (lower.includes("libertadores")) return "Copa Libertadores";
  if (lower.includes("copa do brasil") || lower.includes("brazilian cup")) return "Copa do Brasil";
  if (lower.includes("brasileirao") || lower.includes("brasileirão") || lower.includes("serie a")) return "Brasileirão";
  return name.trim();
}

/**
 * Parse a goal value that may be integer, float, or empty.
 */
export function parseGoal(value: string | null | undefined): number | null {
  if (value === undefined || value === null || value === "") return null;
  const n = Number(value);
  return Number.isFinite(n) ? Math.trunc(n) : null;
}

/**
 * Parse dates in ISO, Brazilian (dd/mm/yyyy), or datetime formats.
 *
 * Returns a UTC Date set to midnight of the calendar date so that dates from
 * different datasets (some with times, some without) line up for deduplication.
 */
export function parseDate(value: string | null | undefined): Date | null {
  if (!value) return null;
  const trimmed = String(value).trim();
  if (!trimmed) return null;

  // ISO / SQL datetime: 2012-05-19 18:30:00 or 2023-09-24
  const isoMatch = trimmed.match(/^(\d{4})-(\d{2})-(\d{2})(?:[ T](\d{2}):(\d{2}):(\d{2}))?/);
  if (isoMatch) {
    const year = Number(isoMatch[1]);
    const month = Number(isoMatch[2]) - 1;
    const day = Number(isoMatch[3]);
    if (day >= 1 && day <= 31 && month >= 0 && month < 12) {
      const d = new Date(Date.UTC(year, month, day));
      if (d.getUTCMonth() === month && d.getUTCDate() === day) return d;
    }
  }

  // Brazilian format: 29/03/2003
  const parts = trimmed.split("/");
  if (parts.length === 3) {
    const day = Number(parts[0]);
    const month = Number(parts[1]) - 1;
    const year = Number(parts[2]);
    if (year > 1900 && year < 2100 && month >= 0 && month < 12 && day >= 1 && day <= 31) {
      const d = new Date(Date.UTC(year, month, day));
      if (d.getUTCMonth() === month && d.getUTCDate() === day) return d;
    }
  }

  return null;
}

/**
 * Parse a season/year value from strings or numbers.
 */
export function parseSeason(value: string | number | null | undefined): number | null {
  if (value === undefined || value === null || value === "") return null;
  const n = typeof value === "number" ? value : Number(value);
  return Number.isFinite(n) && n >= 1900 && n <= 2100 ? n : null;
}

/**
 * Deterministic short ID for a match row (used when no source ID exists).
 */
export function rowId(values: string[]): string {
  return createHash("sha256").update(values.join("|")).digest("hex").slice(0, 16);
}

File: experiments-local/experiment-mu-glm53-provider/runs/agent=oc-zai_language=python_model=openrouter/z-ai/glm-5.3_tooling=none/rep2-failed/normalize.py
"""Normalization utilities for Brazilian soccer data.

The six provided datasets spell club names in wildly different ways:

* ``"Palmeiras-SP"`` / ``"Palmeiras"`` / ``"palmeiras"``
* ``"Athletico Paranaense - PR"`` / ``"Atlético-PR"`` / ``"Athletico"``
* ``"América FC (Minas Gerais)"`` / ``"America - MG"``
* ``"Nacional (URU)"`` / ``"Nacional-URU"``

and dates in at least three formats (ISO, ISO+time, Brazilian DD/MM/YYYY).

This module turns every raw team name into a single *canonical key* so that
matches, players and queries can be joined reliably:

* Brazilian clubs use ``base-uf`` keys (``flamengo-rj``, ``atletico-mg``),
  where the UF comes from an explicit suffix in the data or from the
  built-in alias table below.
* International clubs keep their CONMEBOL country code when present
  (``nacional-uru``, ``barcelona-equ``) or fall back to the bare base
  (``boca juniors``).

It also provides multi-format date parsing, safe integer parsing for goal
columns (``"-"``/``"NA"`` appear in the Libertadores file), competition name
aliases, and the catalog of classic Brazilian derbies.
"""

from __future__ import annotations

import re
import unicodedata
from datetime import date, datetime

UFS = {
    "AC", "AL", "AM", "AP", "BA", "CE", "DF", "ES", "GO", "MA", "MG", "MS",
    "MT", "PA", "PB", "PE", "PI", "PR", "RJ", "RN", "RO", "RR", "RS", "SC",
    "SE", "SP", "TO",
}

COUNTRY_CODES = {
    "ARG", "BOL", "CHI", "COL", "EQU", "MEX", "PAR", "PER", "URU", "VEN",
    "USA", "JPN", "KSA",
}

BASE_ALIASES: dict[str, tuple[str, str | None]] = {
    "flamengo": ("flamengo", "rj"),
    "fluminense": ("fluminense", "rj"),
    "vasco": ("vasco", "rj"),
    "vasco da gama": ("vasco", "rj"),
    "botafogo": ("botafogo", "rj"),
    "botafogo fr": ("botafogo", "rj"),
    "corinthians": ("corinthians", "sp"),
    "sc corinthians paulista": ("corinthians", "sp"),
    "sport club corinthians paulista": ("corinthians", "sp"),
    "palmeiras": ("palmeiras", "sp"),
    "se palmeiras": ("palmeiras", "sp"),
    "palmeiras sp": ("palmeiras", "sp"),
    "sao paulo": ("sao paulo", "sp"),
    "sao paulo fc": ("sao paulo", "sp"),
    "santos": ("santos", "sp"),
    "santos fc": ("santos", "sp"),
    "atletico mineiro": ("atletico", "mg"),
    "atlético mineiro": ("atletico", "mg"),
    "cruzeiro": ("cruzeiro", "mg"),
    "cruzeiro ec": ("cruzeiro", "mg"),
    "america fc": ("america", "mg"),
    "america mineiro": ("america", "mg"),
    "america mg": ("america", "mg"),
    "america de belo horizonte": ("america", "mg"),
    "america fc natal": ("america", "rn"),
    "atletico paranaense": ("atletico", "pr"),
    "athletico paranaense": ("atletico", "pr"),
    "athletico": ("atletico", "pr"),
    "club atletico paranaense": ("atletico", "pr"),
    "atletico goianiense": ("atletico", "go"),
    "atletico acreano": ("atletico", "ac"),
    "atletico alagoinhas": ("atletico", "ba"),
    "atletico cearense": ("atletico cearense", "ce"),
    "fc atletico cearense": ("atletico cearense", "ce"),
    "gremio": ("gremio", "rs"),
    "internacional": ("internacional", "rs"),
    "internacional sc": ("internacional", "rs"),
    "gremio fbpa": ("gremio", "rs"),
    "sport": ("sport", "pe"),
    "sport club do recife": ("sport", "pe"),
    "sport do recife": ("sport", "pe"),
    "sport recife": ("sport", "pe"),
    "santa cruz": ("santa cruz", "pe"),
    "santa cruz fc": ("santa cruz", "pe"),
    "nautico": ("nautico", "pe"),
    "nautico capibaribe": ("nautico", "pe"),
    "bahia": ("bahia", "ba"),
    "ec bahia": ("bahia", "ba"),
    "esporte clube bahia": ("bahia", "ba"),
    "vitoria": ("vitoria", "ba"),
    "ec vitoria": ("vitoria", "ba"),
    "vitoria ec": ("vitoria", "ba"),
    "esporte clube vitoria": ("vitoria", "ba"),
    "vitoria f c": ("vitoria", "es"),
    "vitoria fc": ("vitoria", "es"),
    "fortaleza": ("fortaleza", "ce"),
    "fortaleza ec": ("fortaleza", "ce"),
    "fortaleza fc": ("fortaleza", "ce"),
    "ceara": ("ceara", "ce"),
    "ceara sporting club": ("ceara", "ce"),
    "ceara sc": ("ceara", "ce"),
    "goias": ("goias", "go"),
    "goias ec": ("goias", "go"),
    "avai": ("avai", "sc"),
    "avai fc": ("avai", "sc"),
    "chapecoense": ("chapecoense", "sc"),
    "figueirense": ("figueirense", "sc"),
    "criciuma": ("criciuma", "sc"),
    "criciuma ec": ("criciuma", "sc"),
    "joinville": ("joinville", "sc"),
    "parana": ("parana", "pr"),
    "parana clube": ("parana", "pr"),
    "ca parana": ("parana", "pr"),
    "coritiba": ("coritiba", "pr"),
    "coritiba fc": ("coritiba", "pr"),
    "ponte preta": ("ponte preta", "sp"),
    "aa ponte preta": ("ponte preta", "sp"),
    "portuguesa": ("portuguesa", "sp"),
    "portuguesa sp": ("portuguesa", "sp"),
    "juventude": ("juventude", "rs"),
    "ec juventude": ("juventude", "rs"),
    "cuiaba": ("cuiaba", "mt"),
    "cuiaba ec": ("cuiaba", "mt"),
    "red bull bragantino": ("red bull bragantino", "sp"),
    "bragantino": ("red bull bragantino", "sp"),
    "csa": ("csa", "al"),
    "cs alagoano": ("csa", "al"),
    "centro sportivo alagoano": ("csa", "al"),
    "guarani": ("guarani", "sp"),
    "guarani fc": ("guarani", "sp"),
    "brasiliense": ("brasiliense", "df"),
    "ipatinga": ("ipatinga", "mg"),
    "barueri": ("barueri", "sp"),
    "gremio prudente": ("gremio prudente", "sp"),
    "america de natal": ("america", "rn"),
    "america rn": ("america", "rn"),
    "abc": ("abc", "rn"),
    "atletico madrid": ("atletico madrid", None),
    "athletico paranaense - pr": ("atletico", "pr"),
    "flamengo do piaui": ("flamengo", "pi"),
    "sao jose poa": ("sao jose", "rs"),
    "remo": ("remo", "pa"),
    "clube do remo": ("remo", "pa"),
    "paysandu": ("paysandu", "pa"),
    "tuna luso": ("tuna luso", "pa"),
    "portuguesa desportos": ("portuguesa", "sp"),
    "ferroviario": ("ferroviario", "ce"),
    "caxias": ("caxias", "rs"),
    "ser caxias": ("caxias", "rs"),
    "brasilia fc": ("brasilia", "df"),
    "vila nova": ("vila nova", "go"),
    "crb": ("crb", "al"),
    "asa": ("asa", "al"),
    "a s a": ("asa", "al"),
    "confianca": ("confianca", "se"),
    "ad confianca": ("confianca", "se"),
    "sampaio correa": ("sampaio correa", "ma"),
    "novorizontino": ("gremio novorizontino", "sp"),
    "gremio novorizontino": ("gremio novorizontino", "sp"),
    "ituano": ("ituano", "sp"),
    "santo andre": ("santo andre", "sp"),
    "sao caetano": ("sao caetano", "sp"),
    "mirassol": ("mirassol", "sp"),
    "oeste": ("oeste", "sp"),
    "londrina": ("londrina", "pr"),
    "maringa": ("maringa", "pr"),
    "cianorte": ("cianorte", "pr"),
    "tombense": ("tombense", "mg"),
    "tupi": ("tupi", "mg"),
    "uberlandia": ("uberlandia", "mg"),
    "boa": ("boa", "mg"),
    "anapolina": ("anapolina", "go"),
    "crac": ("crac", "go"),
    "goianesia": ("goianesia", "go"),
    "gama": ("gama", "df"),
    "se gama": ("gama", "df"),
    "bangu": ("bangu", "rj"),
    "nova iguacu": ("nova iguacu", "rj"),
    "volta redonda": ("volta redonda", "rj"),
    "madureira ec": ("madureira", "rj"),
    "boavista": ("boavista", "rj"),
    "boavista sport club": ("boavista", "rj"),
    "boavista sc saquarema": ("boavista", "rj"),
    "cabofriense": ("cabofriense", "rj"),
    "duque de caxias fc": ("duque de caxias", "rj"),
    "americano": ("americano", "rj"),
    "salgueiro": ("salgueiro", "pe"),
    "juazeirense": ("juazeirense", "ba"),
    "bahia de feira": ("bahia de feira", "ba"),
    "imperatriz": ("imperatriz", "ma"),
    "moto clube": ("moto club", "ma"),
    "moto club de sao luis": ("moto club", "ma"),
    "amazonas": ("amazonas", "am"),
    "manaus": ("manaus", "am"),
    "abc fc": ("abc", "rn"),
    "river": ("river", "pi"),
    "retro fc brasil": ("retro", "pe"),
    "guarany de sobral": ("guarany", "ce"),
    "afogados da ingazeira fc": ("afogados", "pe"),
    "esportivo bento goncalves": ("esportivo", "rs"),
    "ceo varzeagrandense": ("operario", "mt"),
    "peixe da amazonia": ("santos", "ap"),
    "brasil de pelotas": ("brasil", "rs"),
    "uniao": ("uniao de rondonopolis", "mt"),
    "inter de limeira": ("inter de limeira", "sp"),
    "sao bento": ("sao bento", "sp"),
    "sao bernardo": ("sao bernardo", "sp"),
    "sao jose rs": ("sao jose", "rs"),
    "macae": ("macae", "rj"),
    "macae esporte fc": ("macae", "rj"),
    "guaratingueta": ("guaratingueta", "sp"),
    "sao luiz": ("sao luiz", "rs"),
    "juventude ma": ("juventude", "ma"),
}

DISPLAY_NAMES: dict[str, str] = {
    "flamengo-rj": "Flamengo",
    "fluminense-rj": "Fluminense",
    "vasco-rj": "Vasco da Gama",
    "botafogo-rj": "Botafogo",
    "corinthians-sp": "Corinthians",
    "palmeiras-sp": "Palmeiras",
    "sao paulo-sp": "São Paulo",
    "santos-sp": "Santos",
    "atletico-mg": "Atlético Mineiro",
    "cruzeiro-mg": "Cruzeiro",
    "america-mg": "América Mineiro",
    "atletico-pr": "Athletico Paranaense",
    "gremio-rs": "Grêmio",
    "internacional-rs": "Internacional",
    "sport-pe": "Sport Recife",
    "santa cruz-pe": "Santa Cruz",
    "nautico-pe": "Náutico",
    "bahia-ba": "Bahia",
    "vitoria-ba": "Vitória",
    "fortaleza-ce": "Fortaleza",
    "ceara-ce": "Ceará",
    "goias-go": "Goiás",
    "avai-sc": "Avaí",
    "chapecoense-sc": "Chapecoense",
    "figueirense-sc": "Figueirense",
    "criciuma-sc": "Criciúma",
    "joinville-sc": "Joinville",
    "parana-pr": "Paraná",
    "coritiba-pr": "Coritiba",
    "ponte preta-sp": "Ponte Preta",
    "portuguesa-sp": "Portuguesa",
    "juventude-rs": "Juventude",
    "cuiaba-mt": "Cuiabá",
    "red bull bragantino-sp": "Red Bull Bragantino",
    "csa-al": "CSA",
    "guarani-sp": "Guarani",
    "brasiliense-df": "Brasiliense",
    "ipatinga-mg": "Ipatinga",
    "baruleri-sp": "Barueri",
    "barueri-sp": "Barueri",
    "gremio prudente-sp": "Grêmio Prudente",
    "america-rn": "América-RN",
    "abc-rn": "ABC",
    "brasil-rs": "Brasil de Pelotas",
}

COMPETITIONS = {
    "brasileirao serie a": "Brasileirão Série A",
    "serie a": "Brasileirão Série A",
    "brasileirao": "Brasileirão Série A",
    "brasileirão": "Brasileirão Série A",
    "campeonato brasileiro": "Brasileirão Série A",
    "campeonato brasileiro serie a": "Brasileirão Série A",
    "serie b": "Brasileirão Série B",
    "brasileirao serie b": "Brasileirão Série B",
    "serie c": "Brasileirão Série C",
    "brasileirao serie c": "Brasileirão Série C",
    "copa do brasil": "Copa do Brasil",
    "brazilian cup": "Copa do Brasil",
    "copa libertadores": "Copa Libertadores",
    "libertadores": "Copa Libertadores",
    "copa libertadores da america": "Copa Libertadores",
}

DERBY_PAIRS: list[tuple[str, str, str]] = [
    ("flamengo-rj", "fluminense-rj", "Fla-Flu"),
    ("flamengo-rj", "vasco-rj", "Clássico dos Milhões"),
    ("botafogo-rj", "fluminense-rj", "Clássico Vovô"),
    ("botafogo-rj", "vasco-rj", "Clássico dos Rivais"),
    ("corinthians-sp", "palmeiras-sp", "Derby Paulista"),
    ("corinthians-sp", "sao paulo-sp", "Majestoso"),
    ("palmeiras-sp", "sao paulo-sp", "Choque-Rei"),
    ("corinthians-sp", "santos-sp", "Clássico Alvinegro"),
    ("sao paulo-sp", "santos-sp", "Clássico San-São"),
    ("gremio-rs", "internacional-rs", "Gre-Nal"),
    ("atletico-mg", "cruzeiro-mg", "Clássico Mineiro"),
    ("bahia-ba", "vitoria-ba", "Ba-Vi"),
    ("atletico-pr", "coritiba-pr", "Atle-Tiba"),
    ("sport-pe", "nautico-pe", "Clássico dos Clássicos"),
    ("sport-pe", "santa cruz-pe", "Clássico das Multidões"),
    ("ceara-ce", "fortaleza-ce", "Clássico-Rei"),
]

_PARENTHESES_RE = re.compile(r"\(([^)]*)\)")
_TRAILING_DASH_RE = re.compile(r"\s*[-–]\s*([a-z]{2,3})\s*$")
_TRAILING_WORD_RE = re.compile(r"\s+([a-z]{2,3})\s*$")
_NON_ALPHA_RE = re.compile(r"[^a-z0-9\s]")


def strip_accents(text: str) -> str:
    """Return *text* with accents removed (NFD decomposition)."""
    decomposed = unicodedata.normalize("NFD", text)
    return "".join(ch for ch in decomposed if not unicodedata.combining(ch))


def _clean_base(text: str) -> str:
    """Lowercase, de-accent, strip punctuation and collapse whitespace."""
    text = strip_accents(text).lower().strip()
    text = _NON_ALPHA_RE.sub(" ", text)
    text = re.sub(r"\s+", " ", text).strip()
    tokens = text.split(" ")
    if len(tokens) > 1 and all(len(t) == 1 for t in tokens):
        text = "".join(tokens)
    return text


def canonical_team_key(raw_name: str) -> str:
    """Convert any raw team spelling into the canonical team key."""
    if not raw_name or not raw_name.strip():
        return ""
    cleaned = strip_accents(raw_name).lower().strip()

    country: str | None = None
    for inside in _PARENTHESES_RE.findall(cleaned):
        token = inside.strip().strip(".")
        if token.upper() in COUNTRY_CODES:
            country = token.upper()
    cleaned = _PARENTHESES_RE.sub(" ", cleaned)

    base = cleaned.strip()
    uf: str | None = None

    dash_match = _TRAILING_DASH_RE.search(base)
    if dash_match:
        suffix = dash_match.group(1).upper()
        if suffix in UFS:
            uf = suffix
            base = base[: dash_match.start()].strip()
        elif suffix in COUNTRY_CODES:
            country = suffix
            base = base[: dash_match.start()].strip()
    if uf is None:
        word_match = _TRAILING_WORD_RE.search(base)
        if word_match and word_match.group(1).upper() in UFS:
            uf = word_match.group(1).upper()
            base = base[: word_match.start()].strip()

    base = _clean_base(base)
    if not base:
        return ""

    if base in BASE_ALIASES:
        canonical_base, default_uf = BASE_ALIASES[base]
        if uf and default_uf and uf.lower() != default_uf:
            pass
        else:
            base = canonical_base
            uf = uf or default_uf

    if uf:
        return f"{base}-{uf.lower()}"
    if country:
        return f"{base}-{country.lower()}"
    return base


def team_display_name(key: str) -> str:
    """Human-friendly display name for a canonical team key."""
    if key in DISPLAY_NAMES:
        return DISPLAY_NAMES[key]
    if "-" in key:
        base, suffix = key.rsplit("-", 1)
        pretty = " ".join(w.capitalize() for w in base.split())
        return f"{pretty} ({suffix.upper()})"
    return " ".join(w.capitalize() for w in key.split())


def resolve_competition(raw: str | None) -> str | None:
    """Map a user-provided competition string to a canonical competition."""
    if not raw:
        return None
    cleaned = _clean_base(raw).replace("  ", " ").strip()
    direct = COMPETITIONS.get(cleaned)
    if direct:
        return direct
    for alias, canonical in COMPETITIONS.items():
        if alias in cleaned or cleaned in alias:
            return canonical
    return None


def parse_date(raw: str | None) -> date | None:
    """Parse ISO, ISO+time and Brazilian (DD/MM/YYYY) date strings."""
    if not raw:
        return None
    text = raw.strip()
    if text in {"", "NA", "None", "-"}:
        return None
    text = text.split(" ")[0] if "T" not in text else text.split("T")[0]
    for fmt in ("%Y-%m-%d", "%d/%m/%Y", "%Y/%m/%d", "%d-%m-%Y"):
        try:
            return datetime.strptime(text, fmt).date()
        except ValueError:
            continue
    return None


def parse_int(raw) -> int | None:
    """Parse an integer that may arrive as int, '2', '-' or 'NA'."""
    if raw is None:
        return None
    if isinstance(raw, int):
        return raw
    text = str(raw).strip()
    if text in {"", "NA", "-", "None"}:
        return None
    try:
        return int(float(text))
    except ValueError:
        return None


def normalize_round(competition: str, raw: str | None) -> str | None:
    """Human-friendly round label; maps Copa do Brasil numeric rounds."""
    value = parse_int(raw)
    if value is None:
        return str(raw).strip() if raw else None
    if competition == "Copa do Brasil":
        names = {
            5: "Round of 16",
            6: "Quarterfinal",
            7: "Semifinal",
            8: "Final",
        }
        return names.get(value, f"Round {value}")
    return f"Round {value}"
